### PR TITLE
slim trie add proto.marshaler and proto.unmarshaler interface

### LINKED ISSUE #82: marshaling support for SlimTrie
**Support proto.Marshal(SlimTrie)**

SlimTrie should implement two interface: `proto.Marshaler` and `proto.Unmarhsaler`.
```go
type Marshaler interface {
	Marshal() ([]byte, error)
}
type Unmarshaler interface {
	Unmarshal([]byte) error
}
```
To let the upper level apps easily marshal/unmarshal a SlimTrie, just like operating on a standard `proto.Message`.

To utilize `protobuf` functionalities.

**API changes**:

Currently SlimTrie provides Marshal and Unmarshal but the return value are different: `func (st *SlimTrie) marshal(writer io.Writer) (cnt int64, err error)`

**On-disk data structure changes**:

Currently on disk format is:
```
Children-header
Children-array
Steps-header
Steps-array
Leaves-header
Leaves-array
```

The `header` is written by package `serialize`.

After adapting to `proto.Marshaler`, the header is not stored any more:
```
Children-array
Steps-array
Leaves-array
```

It should be responsibility of upper level to add the header.
E.g. by calling `serialize.Marshal(slimtrie)` instead of calling `slimtrie.Marshal`.

`serialize.Marshal` writes a header(consists of version and length of the marshaled object), then finaly the on-disk layout should be:
```
SlimTrie-header
Children-array
Steps-array
Leaves-array
```

### PR DESCRIPTION
**Pull request by liubaohai**
# Description

Fixes #82 

## Type of change

- **New feature**     
- **Test changes**

## How to reproduce it (if it is a bug-fix PR)

- Env: x86-64, CentOS-7.4, kernel-3.10.0, GO-1.10.1, etc.

- Step-1:
- Step-2:

## The solution (to fix a bug, implement a new feature etc.)

# Checklist:

- [x] **Style**:       My code follows the **style guidelines** of this project
- [x] **Self-review**: I have performed a **self-review** of my own code
- [ ] **Comment**:     I have **commented my code**, particularly in hard-to-understand areas
- [ ] **Doc**:         I have made corresponding changes to the **documentation**
- [x] **No-warnings**: My changes generate **no new warnings**
- [x] **Add-test**:    I have added **tests** that prove my fix is effective or that my feature works
- [x] **Pass**:        New and existing **unit tests pass** locally with my changes
- [x] **Dep**:         Any **dependent** changes have been merged and published in downstream modules

### CODE REVIEW

File: trie/slimtrie.go [hunk 1]
```diff
@@ -436,3 +437,26 @@ func (st *SlimTrie) unmarshal(reader io.Reader) error {
 
 	return nil
 }
+
+func (st *SlimTrie) Marshal() ([]byte, error) {
+	var buf []byte
+	writer := bytes.NewBuffer(buf)
+	if _, err := st.marshal(writer); err != nil {
```
Comment (drmingdrmer) [inline trie/slimtrie.go]: st.marshal 直接去掉吧...没用了...为了保留代码才留下的

File: trie/slimtrie.go [hunk 2]
```diff
@@ -436,3 +437,44 @@ func (st *SlimTrie) unmarshal(reader io.Reader) error {
...
+
+	if err := serialize.Unmarshal(reader, &st.Leaves); err != nil {
+		return errors.WithMessage(err, "failed to unmarshal leaves")
+	}
+
+	return nil
+}
+
+func (st *SlimTrie) Reset()         {}
+func (st *SlimTrie) String() string { return "" }
```
Comment (drmingdrmer) [inline trie/slimtrie.go]: 这个...返回点啥东西吧...

File: trie/slimtrie.go [hunk 3]
```diff
@@ -436,3 +437,44 @@ func (st *SlimTrie) unmarshal(reader io.Reader) error {
...
+	}
+
+	if err := serialize.Unmarshal(reader, &st.Leaves); err != nil {
+		return errors.WithMessage(err, "failed to unmarshal leaves")
+	}
+
+	return nil
+}
+
+func (st *SlimTrie) Reset()         {}
```
Comment (drmingdrmer) [inline trie/slimtrie.go]: Reset需要实现一下...把内容都重置回去, 貌似应该调用3个array的Reset.

Comment (drmingdrmer) [reply trie/slimtrie.go]: 所以...这里应该加个测试, Unmarshal 1个对象2次看是否OK.

File: trie/slimtrie.go (hunk 2 above)
Comment (liubaohai) [reply trie/slimtrie.go]: 看proto实现，这个接口不会用到。只有Reset接口会用到。

File: trie/slimtrie.go (hunk 3 above)
Comment (liubaohai) [reply trie/slimtrie.go]: `serialize.Unmarshal里面proto.Unmarshal`的时候会调用Reset。这里没啥用

Comment (drmingdrmer) [reply trie/slimtrie.go]: Reset是Message的一个接口, 应该保持行为符合它的预期, 这里测试能通过只是因为SlimTrie的Unmarshal调用了内部的数组的Unmarshal, Reset应该也能独立工作. 也需要独立测试下Reset能符合预期.

File: trie/slimtrie.go (hunk 2 above)
Comment (drmingdrmer) [reply trie/slimtrie.go]: String一般是用在fmt.Println啥的里面时显示这个变量的信息的. 实现接口就实现的完整点. 要不别人用的时候print一下出个空白也挺奇怪的.

**Pull request merged**